File: app/src/test/java/de/roland_illig/phantomgo/CountingBoardTest.java
package de.roland_illig.phantomgo;

import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class CountingBoardTest {

    @Test
    public void testCountSimple() {
        Board board = new Board(9);
        board.setup(
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .",
                ". . B . . W . . .");

        CountingBoard countingBoard = new CountingBoard(board);

        assertThat(countingBoard.regionsToString(), is(""
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"
                + "1 1 2 3 3 4 5 5 5\n"));

        CountResult result = countingBoard.count();
        assertThat(result.blackTerritory, is(18));
        assertThat(result.whiteTerritory, is(27));
        assertThat(result.blackCaptured, is(0));
        assertThat(result.whiteCaptured, is(0));
    }

    @Test
    public void testCountBeforeMarkingDeadStones() {
        Board board = new Board(9);
        board.setup(
                "B . . . W W B . .",
                "B W . W W B B . .",
                "B W . W B B B . .",
                "W W W W W B . . .",
                "B B W B B B . . .",
                ". B B W W B B B B",
                "B B W . W W B W W",
                ". B W W . W W B .",
                ". B W . . . W B .");

        CountingBoard countingBoard = new CountingBoard(board);

        assertThat(countingBoard.regionsToString(), is(""
                + "1 2 2 2 3 3 4 5 5\n"
                + "1 3 2 3 3 4 4 5 5\n"
                + "1 3 2 3 4 4 4 5 5\n"
                + "3 3 3 3 3 4 5 5 5\n"
                + "6 6 3 4 4 4 5 5 5\n"
                + "7 6 6 8 8 4 4 4 4\n"
                + "6 6 9 A 8 8 4 B B\n"
                + "C 6 9 9 D 8 8 E F\n"
                + "C 6 9 D D D 8 E F\n"));

        CountResult result = countingBoard.count();
        assertThat(result.toString(), is("black=15+0, white=5+0"));
    }

    @Test
    public void testCountAfterMarkingDeadStones() {
        Board board = new Board(9);
        board.setup(
                "B . . . W W B . .",
                "B W . W W B B . .",
                "B W . W B B B . .",
                "W W W W W B . . .",
                "B B W B B B . . .",
                ". B B W W B B B B",
                "B B W . W W B W W",
                ". B W W . W W B .",
                ". B W . . . W B .");

        CountingBoard countingBoard = new CountingBoard(board);
        countingBoard.toggleDead(0, 0);
        countingBoard.toggleDead(8, 6);

        CountResult result = countingBoard.count();
        assertThat(result.toString(), is("black=19+2, white=13+3"));

        assertThat(countingBoard.toString(), is(""
                + "# w w w W W B b b\n"
                + "# W w W W B B b b\n"
                + "# W w W B B B b b\n"
                + "W W W W W B b b b\n"
                + "B B W B B B b b b\n"
                + "b B B W W B B B B\n"
                + "B B W w W W B # #\n"
                + "b B W W w W W B b\n"
                + "b B W w w w W B b\n"));
    }

    @Test
    public void testBoardViewAfterMarkingDeadStones() {
        Board board = new Board(9);
        board.setup(
                "B . . . W W B . .",
                "B W . W W B B . .",
                "B W . W B B B . .",
                "W W W W W B . . .",
                "B B W B B B . . .",
                ". B B W W B B B B",
                "B B W . W W B W W",
                ". B W W . W W B .",
                ". B W . . . W B .");

        CountingBoard countingBoard = new CountingBoard(board);
        countingBoard.toggleDead(0, 0);
        countingBoard.toggleDead(8, 6);

        CountResult result = countingBoard.count();
        assertThat(result.toString(), is("black=19+2, white=13+3"));
    }

    @Test
    public void testToggleDead() {
        Board board = new Board(9);
        board.setup(
                "W . . B . . . . W",
                ". . B . . . . W .",
                ". B . . . . W . .",
                "B . . . . W . . .",
                ". . . . W . . . .",
                ". . . W . . . . B",
                ". . W . . . . B .",
                ". W . . . . B . .",
                "W . . . . B . . W");
        CountingBoard countingBoard = new CountingBoard(board);

        countingBoard.toggleDead(0, 8);

        assertThat(countingBoard.toString(), is(""
                + "W . . B b b b b #\n"
                + ". . B b b b b # b\n"
                + ". B b b b b # b b\n"
                + "B b b b b # b b b\n"
                + "b b b b # b b b b\n"
                + "b b b # b b b b B\n"
                + "b b # b b b b B .\n"
                + "b # b b b b B . .\n"
                + "# b b b b B . . W\n"));

        countingBoard.toggleDead(0, 8);

        assertThat(countingBoard.toString(), is(""
                + "W . . B . . . . W\n"
                + ". . B . . . . W .\n"
                + ". B . . . . W . .\n"
                + "B . . . . W . . .\n"
                + ". . . . W . . . .\n"
                + ". . . W . . . . B\n"
                + ". . W . . . . B .\n"
                + ". W . . . . B . .\n"
                + "W . . . . B . . W\n"));
    }
}

File: app/src/test/java/de/roland_illig/phantomgo/CountingBoardTest.kt
package de.roland_illig.phantomgo

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class CountingBoardTest {

    @Test
    fun testCountSimple() {
        val board = Board(9)
        board.setup(
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +",
                "+ + X + + O + + +")

        val countingBoard = CountingBoard(board)

        assertThat(countingBoard.toString()).isEqualTo(""
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n"
                + "x x X + + O o o o\n")

        val result = countingBoard.count()
        assertThat(result.blackTerritory).isEqualTo(18)
        assertThat(result.whiteTerritory).isEqualTo(27)
        assertThat(result.blackCaptured).isEqualTo(0)
        assertThat(result.whiteCaptured).isEqualTo(0)
    }

    @Test
    fun testCountBeforeMarkingDeadStones() {
        val board = Board(9)
        board.setup(
                "X + + + O O X + +",
                "X O + O O X X + +",
                "X O + O X X X + +",
                "O O O O O X + + +",
                "X X O X X X + + +",
                "+ X X O O X X X X",
                "X X O + O O X O O",
                "+ X O O + O O X +",
                "+ X O + + + O X +")

        val countingBoard = CountingBoard(board)

        assertThat(countingBoard.toString()).isEqualTo(""
                + "X + + + O O X x x\n"
                + "X O + O O X X x x\n"
                + "X O + O X X X x x\n"
                + "O O O O O X x x x\n"
                + "X X O X X X x x x\n"
                + "x X X O O X X X X\n"
                + "X X O o O O X O O\n"
                + "x X O O o O O X +\n"
                + "x X O o o o O X +\n")

        val result = countingBoard.count()
        assertThat(result.toString()).isEqualTo("black=15+0, white=5+0")
    }

    @Test
    fun testCountAfterMarkingDeadStones() {
        val board = Board(9)
        board.setup(
                "X + + + O O X + +",
                "X O + O O X X + +",
                "X O + O X X X + +",
                "O O O O O X + + +",
                "X X O X X X + + +",
                "+ X X O O X X X X",
                "X X O + O O X O O",
                "+ X O O + O O X +",
                "+ X O + + + O X +")

        val countingBoard = CountingBoard(board)
        countingBoard.toggleDead(0, 0)
        countingBoard.toggleDead(8, 6)

        val result = countingBoard.count()
        assertThat(result.toString()).isEqualTo("black=19+2, white=13+3")

        assertThat(countingBoard.toString()).isEqualTo(""
                + "# o o o O O X x x\n"
                + "# O o O O X X x x\n"
                + "# O o O X X X x x\n"
                + "O O O O O X x x x\n"
                + "X X O X X X x x x\n"
                + "x X X O O X X X X\n"
                + "X X O o O O X # #\n"
                + "x X O O o O O X x\n"
                + "x X O o o o O X x\n")
    }

    @Test
    fun testBoardViewAfterMarkingDeadStones() {
        val board = Board(9)
        board.setup(
                "X + + + O O X + +",
                "X O + O O X X + +",
                "X O + O X X X + +",
                "O O O O O X + + +",
                "X X O X X X + + +",
                "+ X X O O X X X X",
                "X X O + O O X O O",
                "+ X O O + O O X +",
                "+ X O + + + O X +")

        val countingBoard = CountingBoard(board)
        countingBoard.toggleDead(0, 0)
        countingBoard.toggleDead(8, 6)

        val result = countingBoard.count()
        assertThat(result.toString()).isEqualTo("black=19+2, white=13+3")
    }

    @Test
    fun testToggleDead() {
        val board = Board(9)
        board.setup(
                "O + + X + + + + O",
                "+ + X + + + + O +",
                "+ X + + + + O + +",
                "X + + + + O + + +",
                "+ + + + O + + + +",
                "+ + + O + + + + X",
                "+ + O + + + + X +",
                "+ O + + + + X + +",
                "O + + + + X + + O")
        val countingBoard = CountingBoard(board)

        countingBoard.toggleDead(0, 8)

        assertThat(countingBoard.toString()).isEqualTo(""
                + "O + + X x x x x #\n"
                + "+ + X x x x x # x\n"
                + "+ X x x x x # x x\n"
                + "X x x x x # x x x\n"
                + "x x x x # x x x x\n"
                + "x x x # x x x x X\n"
                + "x x # x x x x X +\n"
                + "x # x x x x X + +\n"
                + "# x x x x X + + O\n")

        countingBoard.toggleDead(0, 8)

        assertThat(countingBoard.toString()).isEqualTo(""
                + "O + + X + + + + O\n"
                + "+ + X + + + + O +\n"
                + "+ X + + + + O + +\n"
                + "X + + + + O + + +\n"
                + "+ + + + O + + + +\n"
                + "+ + + O + + + + X\n"
                + "+ + O + + + + X +\n"
                + "+ O + + + + X + +\n"
                + "O + + + + X + + O\n")
    }
}
